src/Features/Core/Portable/QuickInfo/QuickInfoServiceWithProviders.cs

```diff
@@ -20,46 +21,93 @@ internal abstract class QuickInfoServiceWithProviders : QuickInfoService
-        private ImmutableArray<QuickInfoProvider> _providers;
+        private ImmutableArray<QuickInfoProvider> _externalProviders;
+        private ImmutableArray<InternalQuickInfoProvider> _internalProviders;
 
         protected QuickInfoServiceWithProviders(Workspace workspace, string language)
         {
             _workspace = workspace;
             _language = language;
         }
 
-        private ImmutableArray<QuickInfoProvider> GetProviders()
+        private ImmutableArray<QuickInfoProvider> GetExternalProviders()
         {
-            if (_providers.IsDefault)
+            if (_externalProviders.IsDefault)
             {
-                var mefExporter = (IMefHostExportProvider)_workspace.Services.HostServices;
+                ImmutableInterlocked.InterlockedCompareExchange(ref _externalProviders, CreateProviders<QuickInfoProvider>(), default);
+            }
 
-                var providers = ExtensionOrderer
-                    .Order(mefExporter.GetExports<QuickInfoProvider, QuickInfoProviderMetadata>()
-                        .Where(lz => lz.Metadata.Language == _language))
-                    .Select(lz => lz.Value)
-                    .ToImmutableArray();
+            return _externalProviders;
+        }
 
-                ImmutableInterlocked.InterlockedCompareExchange(ref _providers, providers, default);
+        private ImmutableArray<InternalQuickInfoProvider> GetInternalProviders()
+        {
+            if (_internalProviders.IsDefault)
+            {
+                ImmutableInterlocked.InterlockedCompareExchange(ref _internalProviders, CreateProviders<InternalQuickInfoProvider>(), default);
             }
 
-            return _providers;
+            return _internalProviders;
+        }
+
+        private ImmutableArray<TQuickInfoProvider> CreateProviders<TQuickInfoProvider>()
+        {
+            var mefExporter = (IMefHostExportProvider)_workspace.Services.HostServices;
+
+            return ExtensionOrderer
+                .Order(mefExporter.GetExports<TQuickInfoProvider, QuickInfoProviderMetadata>()
+                    .Where(lz => lz.Metadata.Language == _language))
+                .Select(lz => lz.Value)
+                .ToImmutableArray();
         }
 
         public override async Task<QuickInfoItem?> GetQuickInfoAsync(Document document, int position, CancellationToken cancellationToken)
+        {
+            if (document.SupportsSemanticModel)
+            {
+                var internalContext = await InternalQuickInfoContext.CreateAsync(document, position, cancellationToken).ConfigureAwait(false);
+                var info = await GetQuickInfoAsync(internalContext).ConfigureAwait(false);
+                if (info != null)
+                {
+                    return info;
+                }
+            }
+
+            var context = new QuickInfoContext(document, position, cancellationToken);
+            return await GetQuickInfoAsync(context).ConfigureAwait(false);
```

Comment:
:(  don't hate me.  but this is *not* what i wanted.  specifically, this codified at the *core* layer that there are different types of languages (those that support semantics or not).  What i want is that at the *core* layer that everythign stay the say.  However, the C#/VB particular impl of the core api can then defer to a 'semantic-oriented' api.  LSIF can just bypass the core layer and call into that semantic-oriented api directly.

Reply:
Hmm little confused here - isn't this just an implementation detail? If eventually we decide to make QuickInfoProvider and QuickInfoContext public, we can do so with the current API state in the PR. Whether the implementation detail about semantic model based providers vs document based providers lies in `QuickInfoServiceWithProviders` or another sub-type wouldn't really matter, correct?

Reply:
i'm saying that the C#/VB guys expose multiple ways to call into them.  One is the standard 'i have a doc' way that fits into the langauge-neutral QuickInfo view of hte world.  The second is a new 'i operate at the compilation level' api that clients like LSIF can use instead.

Reply:
I am still not understanding the architectural benefit we would get out of it, given that each of our internal providers will operate on the compilation level. Does embedding this knowledge of dual mode into a single place (`QuickInfoServiceWithProviders`) have any disadvantages over each provider having two APIs and then we replicate the logic from `QuickInfoServiceWithProviders` into LSIF generator to detect relevant providers, iterate through them and invoke the compilation based API? If you don't mind, are you able to push a commit or create a separate PR with your proposed API shape?

Reply:
> are you able to push a commit or create a separate PR with your proposed API shape?

Sure.  let me look at this.